malariagen_data/anoph/hap_freq.py

```diff
@@ -0,0 +1,391 @@
+from typing import Optional  # , Dict, Union, Callable, List
+
+import pandas as pd
+import numpy as np
+import xarray as xr
+import allel
+import dask.array as da
+from numpydoc_decorator import doc  # type: ignore
+
+from ..util import check_types, haplotype_frequencies
+from .hap_data import AnophelesHapData
+from .sample_metadata import locate_cohorts
+from . import base_params, frq_params  # , map_params, plotly_params
+
+
+class AnophelesHapFrequencyAnalysis(
+    AnophelesHapData,
+):
+    def __init__(
+        self,
+        **kwargs,
+    ):
+        # N.B., this class is designed to work cooperatively, and
+        # so it's important that any remaining parameters are passed
+        # to the superclass constructor.
+        super().__init__(**kwargs)
+
+        # Set up cache variables.
+        self._cache_annotator = None
+
+    @check_types
+    @doc(
+        summary="""
+            Compute haplotype frequencies for a region.
+        """,
+        returns="""
+            A dataframe of haplotype frequencies, one row per haplotype.
+        """,
+        notes="""
+            Cohorts with fewer samples than `min_cohort_size` will be excluded from
+            output data frame.
+        """,
+    )
+    def haplotypes_frequencies(
+        self,
+        region: base_params.region,
+        cohorts: base_params.cohorts,
+        sample_query: Optional[base_params.sample_query] = None,
+        min_cohort_size: base_params.min_cohort_size = 10,
+        sample_sets: Optional[base_params.sample_sets] = None,
+        chunks: base_params.chunks = base_params.native_chunks,
+        inline_array: base_params.inline_array = base_params.inline_array_default,
+    ) -> pd.DataFrame:
+        # Access sample metadata.
+        df_samples = self.sample_metadata(
+            sample_sets=sample_sets, sample_query=sample_query
+        )
+
+        # Build cohort dictionary, maps cohort labels to boolean indexers.
+        coh_dict = locate_cohorts(cohorts=cohorts, data=df_samples)
+
+        # Remove cohorts below minimum cohort size.
+        coh_dict = {
+            coh: loc_coh
+            for coh, loc_coh in coh_dict.items()
+            if np.count_nonzero(loc_coh) >= min_cohort_size
+        }
+
+        # Early check for no cohorts.
+        if len(coh_dict) == 0:
+            raise ValueError(
+                "No cohorts available for the given sample selection parameters and minimum cohort size."
+            )
```

Comment:
Feels like all of this could reasonably be moved inside the `locate_cohorts()` function, in which case might also allow removal of code duplicated elsewhere.